### PR TITLE
Nyaml2nxdl comment preservation

### PR DESCRIPTION
**Pull request by RubelMozumder**
The converter `nyaml2nxdl` can now store comments in forward and backward conversion.

One extra class `XMLComment` has already been, albeit it's been not used(instead XML library was exploited), still, it can be used if needed in the future.

### CODE REVIEW

File: nexusutils/nyaml2nxdl/nyaml2nxdl.py [hunk 1]
```diff
@@ -118,32 +116,60 @@ def append_yml(input_file, append, verbose):
...
+        xml_out_file = raw_name + '.nxdl.xml'
+        nyaml2nxdl(input_file, xml_out_file, verbose)
         if append:
-            append_yml(input_file.rsplit(".", 1)[0] + '.nxdl.xml',
+            append_yml(raw_name + '.nxdl.xml',
                        append,
                        verbose
                        )
         else:
             pass
```
Comment (shabihsherjeel) [inline nexusutils/nyaml2nxdl/nyaml2nxdl.py:152]: Is this needed?

File: nexusutils/nyaml2nxdl/nyaml2nxdl_backward_tools.py [hunk 2]
```diff
@@ -30,8 +32,62 @@
...
+
+
+def separate_ip_comments(input_file):
+    comments_list = []
+    comment = []
+    xml_lines = []
+
+    with open(input_file, "r") as file:
+        lines = file.readlines()
+        # pi -> ProcessInstructio section
```
Comment (shabihsherjeel) [inline nexusutils/nyaml2nxdl/nyaml2nxdl_backward_tools.py:45]: Can be removed.

File: nexusutils/nyaml2nxdl/nyaml2nxdl_backward_tools.py [hunk 3]
```diff
@@ -62,6 +118,33 @@ def __init__(
...
+            Parse an XML file provided as input and print a YML file
+        """
+        if os.path.isfile(output_yml):
+            os.remove(output_yml)
+
+        depth = 0
+
+        self.pi_comments, root = parse(input_file)
+        # tree = ET.parse(input_file)
+        # xml_tree = {'tree': tree.getroot(), 'node': tree.getroot()}
```
Comment (shabihsherjeel) [inline nexusutils/nyaml2nxdl/nyaml2nxdl_backward_tools.py:145]: Not needed.

File: nexusutils/nyaml2nxdl/nyaml2nxdl_backward_tools.py [hunk 4]
```diff
@@ -86,11 +185,22 @@ def handle_symbols(self, depth, node):
...
+                            sbl_doc_cmnt_list.append(last_comment)
+                            last_comment = ''
                             self.symbol_list.append(
                                 self.handle_not_root_level_doc(depth,
                                                                tag=child.attrib['name'],
                                                                text=symbol_doc.text))
+        self.strore_root_level_comments('symbol_doc_comments', sbl_doc_cmnt_list)
+        self.strore_root_level_comments('symbol_comments', symbol_cmnt_list)
+
+    def strore_root_level_comments(self, holder, comment):
```
Comment (shabihsherjeel) [inline nexusutils/nyaml2nxdl/nyaml2nxdl_backward_tools.py:200]: typo. "store"

**Review by shabihsherjeel — CHANGES_REQUESTED** [review]

**Review by RubelMozumder — COMMENTED** [review]
Done!

File: nexusutils/nyaml2nxdl/comment_collector.py [hunk 5]
```diff
@@ -0,0 +1,463 @@
...
+        self._next_comment_id = single_comment.cid
+        with open(self.file, mode='r', encoding='UTF-8') as f:
+            lines = f.readlines()
+            # Make an empty line for last comment if no enpty lines in original file
+            lines.append('')
+            for line_num, line in enumerate(lines):
+                if single_comment.is_storing_single_comment():
+                    # If the last comment comes without post nxdl fields, groups and attributes
+                    if line_num < (len(lines) - 1):
+                        # Line number in file starts from 1
```
[the remark this answers is not in the document]
Comment (sanbrock) [reply nexusutils/nyaml2nxdl/comment_collector.py:93]: Well, `line_num` as yielded by `enumerate(lines)` in line 89 normally starts with the value of 0.

**Review by sanbrock — APPROVED** [review]
LGTM

**Pull request merged**